### ISSUE TITLE
aws s3 sync loses track if connection failure: "with -7 file(s) remaining"

### ISSUE THREAD

**Issue by ulsa** [aws/aws-cli#1094]
Lost connection to wifi for a while, during a large sync operation, and got this:

```
upload failed: 20141129/wac_A8AA_20141129_0065.log.gz to s3://my-logs-prod/logs/20141129/wac_A8AA_20141129_0065.log.gz
[Errno 8] _ssl.c:504: EOF occurred in violation of protocol
upload failed: 20141129/wac_A8AA_20141129_0062.log.gz to s3://my-logs-prod/logs/20141129/wac_A8AA_20141129_0062.log.gz
[Errno 8] _ssl.c:504: EOF occurred in violation of protocol
upload failed: 20141129/wac_A8AA_20141129_0066.log.gz to s3://my-logs-prod/logs/20141129/wac_A8AA_20141129_0066.log.gz
[Errno 8] _ssl.c:504: EOF occurred in violation of protocol
upload failed: 20141129/wac_A8AA_20141129_0061.log.gz to s3://my-logs-prod/logs/20141129/wac_A8AA_20141129_0061.log.gz
[Errno 8] _ssl.c:504: EOF occurred in violation of protocol
upload failed: 20141129/wac_A8AA_20141129_0067.log.gz to s3://my-logs-prod/logs/20141129/wac_A8AA_20141129_0067.log.gz
[Errno 8] _ssl.c:504: EOF occurred in violation of protocol
upload failed: 20141129/wac_A8AA_20141129_0068.log.gz to s3://my-logs-prod/logs/20141129/wac_A8AA_20141129_0068.log.gz
[Errno 8] _ssl.c:504: EOF occurred in violation of protocol
upload failed: 20141129/wac_A8AA_20141129_0069.log.gz to s3://my-logs-prod/logs/20141129/wac_A8AA_20141129_0069.log.gz
[Errno 8] _ssl.c:504: EOF occurred in violation of protocol
upload failed: 20141129/wac_A8AA_20141129_0072.log.gz to s3://my-logs-prod/logs/20141129/wac_A8AA_20141129_0072.log.gz [Errno 8] _ssl.c:504: EOF occurred in violation of protocol
upload: 20141129/wac_A8AA_20141129_0070.log.gz to s3://my-logs-prod/logs/20141129/wac_A8AA_20141129_0070.log.gz
upload: 20141129/wac_A8AA_20141129_0071.log.gz to s3://my-logs-prod/logs/20141129/wac_A8AA_20141129_0071.log.gz
Completed 699 of 770 part(s) with -7 file(s) remaining
```

```
$ aws --version
aws-cli/1.6.10 Python/2.7.5 Darwin/13.4.0
```

**Comment by jamesls** [comment]
I'll need to confirm this, but searching around, this might be related to http://stackoverflow.com/questions/14102416/python-requests-requests-exceptions-sslerror-errno-8-ssl-c504-eof-occurred.  We'll need to upgrade our underlying http library to fix this.

**Comment by JordonPhillips** [comment]
I did some digging and found that the issue is related to SNI support, which is not available in the python standard library for Python 2.x. SNI support for 2.x is in pyopenssl, which is part of urllib3, which is the underlying library that requests uses. requests will use pyopenssl only if it is available. When we [upgraded to requests 2.7.0](https://github.com/boto/botocore/commit/a2166c17e0da4c8faa0d677d637d182bb0de8176), we vendored in pyopenssl. This issue should, therefore, be fixed.
